uniffi_bindgen/src/bindings/gecko/gen_gecko.rs

```diff
@@ -0,0 +1,355 @@
+    config: &'config Config,
+    ci: &'ci ComponentInterface,
+    functions: &'functions [Function],
+}
+
+impl<'config, 'ci, 'functions> Namespace<'config, 'ci, 'functions> {
+    pub fn new(
+        config: &'config Config,
+        ci: &'ci ComponentInterface,
+        functions: &'functions [Function],
+    ) -> Self {
+        Self {
+            config: config,
+            ci,
+            functions,
+        }
+    }
+}
+
+/// A header file generated for an interface.
+#[derive(Template)]
+#[template(syntax = "c", escape = "none", path = "InterfaceHeaderTemplate.h")]
+pub struct InterfaceHeader<'config, 'ci, 'obj> {
+    config: &'config Config,
+    ci: &'ci ComponentInterface,
+    obj: &'obj Object,
+}
+
+impl<'config, 'ci, 'obj> InterfaceHeader<'config, 'ci, 'obj> {
+    pub fn new(config: &'config Config, ci: &'ci ComponentInterface, obj: &'obj Object) -> Self {
+        Self {
+            config: config,
+            ci,
+            obj,
+        }
+    }
+}
+
+/// An implementation file generated for a namespace with top-level functions.
+#[derive(Template)]
+#[template(syntax = "cpp", escape = "none", path = "InterfaceTemplate.cpp")]
+pub struct Interface<'config, 'ci, 'obj> {
+    config: &'config Config,
+    ci: &'ci ComponentInterface,
+    obj: &'obj Object,
+}
+
+impl<'config, 'ci, 'obj> Interface<'config, 'ci, 'obj> {
+    pub fn new(config: &'config Config, ci: &'ci ComponentInterface, obj: &'obj Object) -> Self {
+        Self {
+            config: config,
+            ci,
+            obj,
+        }
+    }
+}
+
+/// Filters for our Askama templates above. These output C++, XPIDL, and
+/// WebIDL.
+mod filters {
+    use super::*;
+    use std::fmt;
+
+    /// Declares a WebIDL type in the interface for this library.
+    pub fn type_webidl(type_: &Type) -> Result<String, askama::Error> {
+        Ok(match type_ {
+            Type::Int8 => "byte".into(),
+            Type::UInt8 => "octet".into(),
+            Type::Int16 => "short".into(),
+            Type::UInt16 => "unsigned short".into(),
+            Type::Int32 => "long".into(),
+            Type::UInt32 => "unsigned long".into(),
+            Type::Int64 => "long long".into(),
+            Type::UInt64 => "unsigned long long".into(),
+            Type::Float32 => "float".into(),
+            // Note: Not `unrestricted double`; we don't want to allow NaNs
+            // and infinity.
+            Type::Float64 => "double".into(),
+            Type::Boolean => "boolean".into(),
+            Type::String => "DOMString".into(),
+            Type::Enum(name) | Type::Record(name) | Type::Object(name) => class_name_webidl(name)?,
+            Type::Error(name) => panic!("[TODO: type_webidl({:?})]", type_),
+            Type::Optional(type_) => format!("{}?", type_webidl(type_)?),
+            Type::Sequence(type_) => format!("sequence<{}>", type_webidl(type_)?),
+            Type::Map(type_) => format!("record<DOMString, {}>", type_webidl(type_)?),
```

Comment:
Terminology clarification that probably ought to go in a comment: what UniFFI calls a “record”, Firefox calls a “dictionary”. What UniFFI calls a “map”, Firefox calls a “record”. What UniFFI calls an “optional”, Firefox calls “nullable” (“optional” means something different in WebIDL/Firefox).